### PR TITLE
feat(aws): add support for resizing ASGs with optional min/max/desired

### PR DESCRIPTION
**Pull request by dreynaud**
We need to replace ResizeAsgDescription.Capacity with something that has
optional (i.e. nullable) fields. ServerGroup.Capacity happens to have
exactly that.

### CODE REVIEW

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy [hunk 1]
```diff
@@ -110,8 +109,38 @@ class ResizeAsgAtomicOperationUnitSpec extends Specification {
     capacity(0, 2, 3)    | 1          | 2          | 3              || true
   }
 
+  void "should support resize requests with only some fields to update"() {
+    def mockAutoScaling = Mock(AmazonAutoScaling)
```
Comment (ajordens) [inline clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy]: This should be in a `given:` block (or `setup:` as this specification appears to be using)

Looks like this setup is duplicated a bit, could extract to class-level:

```
class ResizeAsgAtomicOperationUnitSpec extends Specification {

  def amazonAutoScaling = Mock(AmazonAutoScaling)
  def amazonClientProvider = Mock(AmazonClientProvider) {
    _ * getAutoScaling(_, _, true) >> { return amazonAutoScaling }
  }

  def setupSpec() {
    TaskRepository.threadLocalTask.set(Mock(Task))
  }

  ...
```

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy [hunk 2]
```diff
@@ -110,8 +109,38 @@ class ResizeAsgAtomicOperationUnitSpec extends Specification {
...
+    def mockAmazonClientProvider = Mock(AmazonClientProvider)
+    mockAmazonClientProvider.getAutoScaling(_, _, true) >> mockAutoScaling
+    mockAutoScaling.describeAutoScalingGroups(_) >> new DescribeAutoScalingGroupsResult().withAutoScalingGroups(
```
Comment (ajordens) [inline clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy]: I'd probably put this in the `then:` block

```
1 * mockAutoScaling.describeAutoScalingGroups(_) >> { ... }
```

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy [hunk 3]
```diff
@@ -110,8 +109,38 @@ class ResizeAsgAtomicOperationUnitSpec extends Specification {
...
+      new AutoScalingGroup().withAutoScalingGroupName("myasg-stack-v0001")
+    )
+
+    def description = new ResizeAsgDescription(
+      asgs: [[
+               serverGroupName: "myasg-stack-v000",
+               region         : "us-west-1",
+               capacity       : new ServerGroup.Capacity(min: 42) // no max or desired specified
```
Comment (ajordens) [inline clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy]: I would suggest parameterizing this capacity and then using a `where:` to pass in a variety of combinations

```
sourceMin | sourceMax | sourceDesired || expectedMin || expectedMax || expectedDesired
42        | null      | null          || 42          || null        || null
42        | 40        | null          || 42          || null        || null

... etc.
```

The split between inputs and outputs might not be necessary (since they're expected to be identical) but you get the point.

Motivation being that we're actually asserting that it works for all combinations.

Comment (dreynaud) [reply clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy]: good idea. What do you think should happen when we have `(min: null, max: null, desired: null)`?

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy (hunk 2 above)
Comment (dreynaud) [reply clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy]: done in 6fbc3d7d7

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy (hunk 1 above)
Comment (dreynaud) [reply clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy]: good point, done in 6fbc3d7d7

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperation.groovy [hunk 4]
```diff
@@ -57,9 +58,15 @@ class ResizeAsgAtomicOperation implements AtomicOperation<Void> {
 
   private void resizeAsg(String asgName,
                          String region,
-                         ResizeAsgDescription.Capacity capacity,
+                         ServerGroup.Capacity capacity,
                          ResizeAsgDescription.Constraints constraints) {
     task.updateStatus PHASE, "Beginning resize of ${asgName} in ${region} to ${capacity}."
+
+    if (capacity.min == null && capacity.max == null && capacity.desired == null) {
+      task.updateStatus PHASE, "Skipping resize of ${asgName} in ${region}, at least one field in ${capacity} needs to be non-null"
```
Comment (ajordens) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperation.groovy]: 👍

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy [hunk 5]
```diff
@@ -110,8 +111,65 @@ class ResizeAsgAtomicOperationUnitSpec extends Specification {
...
   }
 
+  @Unroll
+  void "should support resize requests with only some fields to update (min: #minSize, max: #maxSize, desired: #desiredCapacity)"() {
+    setup:
+    def description = new ResizeAsgDescription(
+      asgs: [[
+               serverGroupName: "myasg-stack-v000",
+               region         : "us-west-1",
+               capacity       : new ServerGroup.Capacity(min: minSize, max: maxSize, desired: desiredCapacity)
```
Comment (ajordens) [inline clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy]: IMO I'd probably be consistent on the naming ... minSize vs maxSize vs desiredCapacity.

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy [hunk 6]
```diff
@@ -110,8 +111,65 @@ class ResizeAsgAtomicOperationUnitSpec extends Specification {
...
+    where:
+    minSize | maxSize | desiredCapacity
+    42      | null    | null
+    null    | 42      | null
+    null    | null    | 42
+    0       | null    | null
+    0       | 0       | 0    // not the same as (null, null, null)
+  }
+
+  void "skip resizing if all capacity fields are null"() {
```
Comment (ajordens) [inline clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy]: I think you could avoid the need for this whole test if you just added a `null | null | null` to the previous test. 

You could add a new column:

```
minSize | maxSize | desiredSize || expectedAutoScalingInvocations 
```

and then

```
expectedAutoScalingInvocations * mockAutoScaling. describeAutoScalingGroups(_) >> {
expectedAutoScalingInvocations * mockAutoScaling.updateAutoScalingGroup(_) >> {
...
}
```

The null/null/null test is certainly valid, this suggestion is just aimed at getting coverage without needing 20 or so additional lines of testing code.

**Comment by ajordens** [comment]
Feel free to merge once the test spec is updated.

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy (hunk 5 above)
Comment (dreynaud) [reply clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy]: Ha, I know. I am sticking to the AWS terminology though

Comment (ajordens) [reply clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy]: Reasonable argument I suppose

Comment (ajordens) [reply clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy]: Reasonable argument I suppose

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy (hunk 6 above)
Comment (dreynaud) [reply clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperationUnitSpec.groovy]: in bde71d874
